### PR TITLE
Cross entropy layer

### PR DESCRIPTION
**Pull request by timmoon10**
This PR implements a layer that takes predicted and ground truth distributions and outputs the cross entropy. This allows us to perform training with all major computation on GPUs, albeit without metrics. This is progress toward #193.

Note that this PR depends on #464, so the diff will be much less intimidating after that is merged.

### CODE REVIEW

File: include/lbann/layers/loss/cross_entropy.hpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+    case data_layout::DATA_PARALLEL:
+      workspace.reset(new StarVCMat<Dev>(prediction.Grid(),
+                                         prediction.Root()));
+      break;
+    case data_layout::MODEL_PARALLEL:
+      workspace.reset(new StarMRMat<Dev>(prediction.Grid(),
+                                         prediction.Root()));
+      break;
+    default: LBANN_ERROR("invalid data layout");
+    }
```
Comment (ndryden) [inline include/lbann/layers/loss/cross_entropy.hpp]: Why do we allocate a distributed matrix every time we do forward prop? That seems expensive.

File: include/lbann/layers/loss/cross_entropy.hpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
```
Comment (ndryden) [inline include/lbann/layers/loss/cross_entropy.hpp]: Why do we allocate a new distributed matrix every time?

File: include/lbann/layers/loss/cross_entropy.hpp (hunk 1 above)
Comment (timmoon10) [reply include/lbann/layers/loss/cross_entropy.hpp]: It's not that bad since we only allocate memory equal to the local mini-batch size. It's less of a problem on GPUs since we use CUB's memory pool. Since the layer deallocates the memory at the end of its computation, it's also (very very barely) more efficient with GPU memory.

**Comment by timmoon10** [comment]
@samadejacobs From the perspective of the layer, there's no way to distinguish predictions or ground truth.

**Comment by timmoon10** [comment]
@ndryden Gradient checks come back clean. I don't notice significant performance differences on AlexNet and Resnet-50 when I disable metrics.

File: model_zoo/models/alexnet/model_alexnet.prototext [hunk 3]
```diff
@@ -56,15 +56,28 @@ model {
...
   layer {
     name: "data"
-    children: "conv1 target"
+    children: "images labels"
     data_layout: "data_parallel"
     input {
       io_buffer: "partitioned"
     }
   }
+  layer {
```
Comment (bvanessen) [inline model_zoo/models/alexnet/model_alexnet.prototext]: Why are there identity layers added here?  It seems like you can just take the data directly to the consuming layer.

File: model_zoo/models/alexnet/model_alexnet.prototext [hunk 4]
```diff
@@ -267,15 +301,28 @@ model {
...
+    name: "cross_entropy_eval"
+    parents: "cross_entropy"
+    data_layout: "model_parallel"
+    evaluation {}    
+  }
   layer {
     name: "target"
-    parents: "prob data"
+    parents: "prob labels"
```
Comment (bvanessen) [inline model_zoo/models/alexnet/model_alexnet.prototext]: Ah, are the identity layers added to allow the implicit split layer to be defined?

File: model_zoo/models/alexnet/model_alexnet.prototext [hunk 5]
```diff
@@ -267,15 +301,28 @@ model {
...
     data_layout: "data_parallel"
     target {}
-  }
+  }
```
Comment (bvanessen) [inline model_zoo/models/alexnet/model_alexnet.prototext]: What is the role of the target layer with the cross entropy layers.

File: include/lbann/layers/loss/cross_entropy.hpp (hunk 1 above)
Comment (bvanessen) [reply include/lbann/layers/loss/cross_entropy.hpp]: Understood that it is not that bad, but why aren't these matrices just setup in the setup_data function?

File: model_zoo/models/alexnet/model_alexnet.prototext [hunk 6]
```diff
@@ -12,7 +12,7 @@ model {
   ###################################################
 
   objective_function {
-    cross_entropy {}
+    layer_term { layer: "cross_entropy_eval" }
```
Comment (bvanessen) [inline model_zoo/models/alexnet/model_alexnet.prototext]: Maybe I just didn't see it in the code review, but how is the layer term used now?

File: model_zoo/models/alexnet/model_alexnet.prototext (hunk 4 above)
Comment (timmoon10) [reply model_zoo/models/alexnet/model_alexnet.prototext]: Yes, perhaps making them split layers would be a little clearer.

File: model_zoo/models/alexnet/model_alexnet.prototext (hunk 5 above)
Comment (timmoon10) [reply model_zoo/models/alexnet/model_alexnet.prototext]: Accuracy is still computed with the target layer. If we don't need to report accuracy, the target layer can be removed.

File: model_zoo/models/alexnet/model_alexnet.prototext (hunk 6 above)
Comment (timmoon10) [reply model_zoo/models/alexnet/model_alexnet.prototext]: The evaluation layer was introduced in #276:
```
objective_function {
  layer_term {
    layer: "name_of_an_evaluation_layer"
    scale_factor: 1.0 // defaults to 1.0
  }
}
```
In the future, I'd like to add some logic to the model class so it can automatically insert evaluation layers where they're needed rather than forcing the practitioner to add them.
